dll/win32/shell32/brfolder.cpp

```diff
@@ -745,50 +735,54 @@ BrFolder_NewFolder(BrFolder *info)
     if (FAILED_UNEXPECTEDLY(hr))
         return hr;
 
-    HTREEITEM hAdded = BrFolder_InsertItem(info, cur, pidlNew, item_data->pidlFull, NULL, hParent);
-
+    HTREEITEM hAdded = BrFolder_InsertItem(info, cur, pidlNew, item_data->pidlFull, hParent);
     TreeView_SortChildren(info->hwndTreeView, hParent, FALSE);
     return BrFolder_Rename(info, hAdded);
 }
 
-static void
-BrFolder_OnCommand(BrFolder *info, UINT id)
+static BOOL
+BrFolder_OnOK(BrFolder *info)
 {
-    LPBROWSEINFOW lpBrowseInfo = info->lpBrowseInfo;
+    // Get the text
     WCHAR szPath[MAX_PATH];
+    GetDlgItemTextW(info->hWnd, IDC_BROWSE_FOR_FOLDER_FOLDER_TEXT, szPath, _countof(szPath));
+    StrTrimW(szPath, L" \t");
+
+    // The original pidl is owned by the treeview and will be free'd.
+    if (!PathIsRelativeW(szPath) && PathIsDirectoryW(szPath))
+        info->pidlRet = ILCreateFromPathW(szPath); // It's valid path
+    else
+        info->pidlRet = ILClone(info->pidlRet);
+
+    if (!info->pidlRet) // A null pidl would mean a cancel
+        info->pidlRet = _ILCreateDesktop();
+
+    pdump(info->pidlRet);
 
+    LPBROWSEINFOW lpBrowseInfo = info->lpBrowseInfo;
+    if (!lpBrowseInfo->pszDisplayName)
+        return TRUE;
+
+    SHFILEINFOW fileInfo = { NULL };
+    lpBrowseInfo->pszDisplayName[0] = UNICODE_NULL;
+    if (SHGetFileInfoW((LPCWSTR)info->pidlRet, 0, &fileInfo, sizeof(fileInfo),
+                       SHGFI_PIDL | SHGFI_DISPLAYNAME))
+    {
+        lstrcpynW(lpBrowseInfo->pszDisplayName, fileInfo.szDisplayName, MAX_PATH);
+    }
+
+    return TRUE;
```

Comment:
Not sure whether your `BrFolder_OnOK` really needs to return a boolean, since it always returns TRUE. You could make the function "return" void instead.

Reply:
Fixed.